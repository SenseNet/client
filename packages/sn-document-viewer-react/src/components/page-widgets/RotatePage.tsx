import { PreviewImageData } from '@sensenet/client-core'
import IconButton from '@material-ui/core/IconButton'
import RotateLeft from '@material-ui/icons/RotateLeft'
import RotateRight from '@material-ui/icons/RotateRight'
import React, { useCallback, useState } from 'react'
import { useLocalization, usePreviewImage } from '../../hooks'
import { Dimensions } from '../../services'

/**
 * The amount of rotation in degrees
 */
export const ROTATION_AMOUNT = 90

/**
 * Visibility of the rotation related buttons
 */
export enum ROTATION_MODE {
  clockwise,
  anticlockwise,
  all,
}

/**
 * Defined the component's own properties
 */
export interface RotatePageProps {
  page: PreviewImageData
  viewPort: Dimensions
  zoomRatio: number
}

export const RotatePageWidget: React.FC<RotatePageProps> = (props) => {
  const localization = useLocalization()

  const [pageIndex] = useState(props.page.Index)
  const image = usePreviewImage(pageIndex)

  const rotatePageLeft = useCallback(() => {
    image.rotate(-ROTATION_AMOUNT)
  }, [image])

  const rotatePageRight = useCallback(() => {
    image.rotate(ROTATION_AMOUNT)
  }, [image])

  return (
    <div
      style={{
        position: 'absolute',
        zIndex: 1,
        top: 0,
        right: 0,
        filter: 'drop-shadow(0 0 3px white) drop-shadow(0 0 5px white) drop-shadow(0 0 9px white)',
      }}>
      <IconButton onClick={rotatePageLeft} title={localization.rotatePageLeft}>
        <RotateLeft />
      </IconButton>

      <IconButton onClick={rotatePageRight} title={localization.rotatePageRight}>
        <RotateRight />
      </IconButton>
    </div>
  )
}
