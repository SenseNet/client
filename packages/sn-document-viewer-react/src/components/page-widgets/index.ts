export * from './RotatePage'
export * from './Shape'
export * from './Shapes'
export * from './style'
