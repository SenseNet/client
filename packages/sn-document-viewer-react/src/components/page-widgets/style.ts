import styled from 'styled-components'
import { DraftCommentMarker } from '../../models'

export const ShapesContainer = styled.div`
  position: absolute;
  top: 0px;
  left: 0px;
  width: 100%;
  height: 100%;
  z-index: 1;
`
export const MARKER_SIZE = 10

export const CommentMarker = styled('div')<{ marker: DraftCommentMarker; zoomRatio: number; isSelected: boolean }>`
  position: absolute;
  top: ${props => props.marker.y * props.zoomRatio}px;
  left: ${props => props.marker.x * props.zoomRatio}px;
  width: ${MARKER_SIZE}px;
  height: ${MARKER_SIZE}px;
  border-radius: ${MARKER_SIZE}px;
  background-color: ${props =>
    props.isSelected ? props.theme.palette.primary.light : props.theme.palette.primary.dark};
  cursor: pointer;
`
