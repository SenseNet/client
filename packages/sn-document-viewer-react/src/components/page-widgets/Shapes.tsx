import { Annotation, Highlight, PreviewImageData, Redaction, Shape, Shapes } from '@sensenet/client-core'
import React, { useCallback } from 'react'
import { useComments, useCommentState, useDocumentData, useDocumentPermissions, useViewerState } from '../../hooks'
import { Dimensions } from '../../services'
import { ShapeSkeleton } from '../shapes'
import { CommentMarker } from './style'

/**
 * Defined the component's own properties
 */
export interface ShapesWidgetProps {
  page: PreviewImageData
  zoomRatio: number
}

/**
 * Page widget component for displaying shapes on a page
 */
export const ShapesWidget: React.FC<ShapesWidgetProps> = (props) => {
  const permissions = useDocumentPermissions()
  const viewerState = useViewerState()
  const { documentData, updateDocumentData } = useDocumentData()
  const comments = useComments()
  const commentState = useCommentState()

  const visibleShapes = {
    redactions: documentData.shapes.redactions.filter((r) => r.imageIndex === props.page.Index) as Redaction[],
    highlights: documentData.shapes.highlights.filter((r) => r.imageIndex === props.page.Index) as Highlight[],
    annotations: documentData.shapes.annotations.filter((r) => r.imageIndex === props.page.Index) as Annotation[],
  }

  const visibleComments = [
    ...comments.comments.filter((comment) => comment.page === props.page.Index),
    ...(props.page.Index !== undefined && commentState.draft?.page === props.page?.Index ? [commentState.draft] : []),
  ]

  const removeShape = useCallback(
    (shapeType: keyof Shapes, guid: string) => {
      ;(documentData.shapes as any)[shapeType] = documentData.shapes[shapeType].filter((s) => s.guid !== guid)
      updateDocumentData(documentData)
      viewerState.updateState({ hasChanges: true })
    },
    [documentData, updateDocumentData, viewerState],
  )

  const updateShapeData = useCallback(
    (shapeType: keyof Shapes, guid: string, shapeChange: Shape | Annotation) => {
      ;(documentData.shapes as any)[shapeType] = (documentData.shapes[shapeType] as Shape[]).map((s) => {
        if (s.guid === guid) {
          return { ...s, ...shapeChange }
        }
        return s
      })
      viewerState.updateState({ hasChanges: true })
      updateDocumentData(documentData)
    },
    [documentData, updateDocumentData, viewerState],
  )

  const onDrop = useCallback(
    (ev: React.DragEvent<HTMLElement>) => {
      if (permissions.canEdit) {
        ev.preventDefault()
        const shapeData = JSON.parse(ev.dataTransfer.getData('shape')) as {
          type: keyof Shapes
          shape: Shape
          offset: Dimensions
        }
        const boundingBox = ev.currentTarget.getBoundingClientRect()
        updateShapeData(shapeData.type, shapeData.shape.guid, {
          ...shapeData.shape,
          imageIndex: props.page.Index,
          x: (ev.clientX - boundingBox.left - shapeData.offset.width) * (1 / props.zoomRatio),
          y: (ev.clientY - boundingBox.top - shapeData.offset.height) * (1 / props.zoomRatio),
        })
      }
    },
    [permissions.canEdit, props.page.Index, props.zoomRatio, updateShapeData],
  )
  return (
    <div
      style={{
        position: 'absolute',
        top: 0,
        left: 0,
        width: '100%',
        height: '100%',
        zIndex: 1,
      }}
      onDrop={onDrop}
      onDragOver={(ev) => ev.preventDefault()}>
      {viewerState.showComments &&
        visibleComments.length > 0 &&
        visibleComments.map((marker) => (
          <CommentMarker
            onClick={() => commentState.setActiveComment(marker.id)}
            isSelected={marker.id === commentState.activeCommentId}
            zoomRatio={props.zoomRatio}
            marker={marker}
            key={marker.id}
          />
        ))}
      <div>
        {permissions.canHideRedaction &&
          viewerState.showRedaction &&
          visibleShapes.redactions.map((redaction, index) => {
            return (
              <ShapeSkeleton
                key={index}
                shape={redaction}
                shapeType="redactions"
                zoomRatio={props.zoomRatio}
                updateShapeData={updateShapeData}
                removeShape={removeShape}
              />
            )
          })}

        {viewerState.showShapes &&
          visibleShapes.annotations.map((annotation, index) => {
            return (
              <ShapeSkeleton
                key={index}
                shape={annotation}
                shapeType="annotations"
                zoomRatio={props.zoomRatio}
                updateShapeData={updateShapeData}
                removeShape={removeShape}
              />
            )
          })}

        {viewerState.showShapes &&
          visibleShapes.highlights.map((highlight, index) => {
            return (
              <ShapeSkeleton
                key={index}
                shape={highlight}
                shapeType="highlights"
                zoomRatio={props.zoomRatio}
                updateShapeData={updateShapeData}
                removeShape={removeShape}
              />
            )
          })}
      </div>
    </div>
  )
}
