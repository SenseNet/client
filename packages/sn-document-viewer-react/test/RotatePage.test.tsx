import RotateLeft from '@material-ui/icons/RotateLeft'
import RotateRight from '@material-ui/icons/RotateRight'
import { shallow } from 'enzyme'
import React from 'react'
import { RotatePageComponent, ROTATION_AMOUNT } from '../src/components/page-widgets/RotatePage'
import { examplePreviewImageData } from './__Mocks__/viewercontext'

describe('RotatePage component', () => {
  const locals = {
    rotateLeft: 'rotatePageLeft',
    rotateRight: 'rotatePageRight',
  }
  it('Should render without crashing', () => {
    const wrapper = shallow(
      <RotatePageComponent
        {...locals}
        page={examplePreviewImageData}
        viewPort={{ width: 1024, height: 768 }}
        zoomRatio={1}
        rotateImages={jest.fn()}
        rotateShapesForPages={jest.fn()}
      />,
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('RotateLeft should trigger a rotate to left', () => {
    const rotateImages = jest.fn()
    const rotateShapesForPages = jest.fn()
    const wrapper = shallow(
      <RotatePageComponent
        {...locals}
        page={examplePreviewImageData}
        viewPort={{ width: 1024, height: 768 }}
        zoomRatio={1}
        rotateImages={rotateImages}
        rotateShapesForPages={rotateShapesForPages}
      />,
    )

    wrapper.find(RotateLeft).simulate('click')
    expect(rotateImages).toBeCalledWith([examplePreviewImageData.Index], -ROTATION_AMOUNT)
    expect(rotateShapesForPages).toBeCalledWith(
      [
        {
          index: examplePreviewImageData.Index,
          size: { width: examplePreviewImageData.Width, height: examplePreviewImageData.Height },
        },
      ],
      -ROTATION_AMOUNT,
    )
  })

  it('RotateRight should trigger a rotate to left', () => {
    const rotateImages = jest.fn()
    const rotateShapesForPages = jest.fn()
    const wrapper = shallow(
      <RotatePageComponent
        {...locals}
        page={examplePreviewImageData}
        viewPort={{ width: 1024, height: 768 }}
        zoomRatio={1}
        rotateImages={rotateImages}
        rotateShapesForPages={rotateShapesForPages}
      />,
    )

    wrapper.find(RotateRight).simulate('click')
    expect(rotateImages).toBeCalledWith([examplePreviewImageData.Index], ROTATION_AMOUNT)
    expect(rotateShapesForPages).toBeCalledWith(
      [
        {
          index: examplePreviewImageData.Index,
          size: { width: examplePreviewImageData.Width, height: examplePreviewImageData.Height },
        },
      ],
      ROTATION_AMOUNT,
    )
  })
})
