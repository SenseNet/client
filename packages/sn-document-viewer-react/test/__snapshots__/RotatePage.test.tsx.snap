// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RotatePage component Should render without crashing 1`] = `
<div
  style={
    Object {
      "filter": "drop-shadow(0 0 3px white) drop-shadow(0 0 5px white) drop-shadow(0 0 9px white)",
      "position": "absolute",
      "right": 0,
      "top": 0,
      "zIndex": 1,
    }
  }
>
  <WithStyles(ForwardRef(IconButton))
    onClick={[Function]}
    title="Rotate page left"
  >
    <RotateLeftIcon />
  </WithStyles(ForwardRef(IconButton))>
  <WithStyles(ForwardRef(IconButton))
    onClick={[Function]}
    title="Rotate page right"
  >
    <RotateRightIcon />
  </WithStyles(ForwardRef(IconButton))>
</div>
`;
